rasa/cdu/command_processor.py

```diff
@@ -155,17 +157,21 @@ def execute_commands(
                 context=context,
             )
 
-            if current_top_flow.id != FLOW_PATTERN_CORRECTION_ID:
+            if top_non_question_frame.flow_id != FLOW_PATTERN_CORRECTION_ID:
                 flow_stack.push(correction_frame)
             else:
                 # wrap up the previous correction flow
-                flow_stack.frames[-1].step_id = END_STEP
+                for i, frame in enumerate(reversed(flow_stack.frames)):
+                    frame.step_id = END_STEP
+                    if frame.frame_id == top_non_question_frame.frame_id:
+                        break
```

Comment:
Is it correct to say that before we break at `top_non_question_frame`, we are adding an END_STEP to question frame? What's the implication of that?

Reply:
we are not really adding a step to the question flow but rather we are setting the stack frame to the end step of that flow. so the step is already there but we "progress" that flow to its end step. 

tl;dr. we remove the question frame from the stack, but in a graceful way where the flow has a chance to wrap up in its end step.